### PR TITLE
Add explicit null check for VSD via tail call stub.

### LINKED ISSUE #17585: [arm32] Assert failure: ExecutionManager::IsManagedCode(adrRedirectedIP)
Consistent failure in Windows arm32 corefx test System.Runtime.Tests with COMPlus_JitStress=2:

https://ci.dot.net/job/dotnet_coreclr/job/master/view/arm/job/jitstress/job/arm_cross_checked_windows_nt_corefx_jitstress2_tst/10/consoleText

```
Running: C:\Users\robox\j\workspace\arm_cross_che---6fedd1e9\_\fx\bin\tests\System.Runtime.Tests\netcoreapp-Windows_NT-Release-arm\RunTests.cmd C:\Users\robox\j\workspace\arm_cross_che---6fedd1e9\_\fx\bin\testhost\netcoreapp-Windows_NT-Release-arm
Using C:\Users\robox\j\workspace\arm_cross_che---6fedd1e9\_\fx\bin\testhost\netcoreapp-Windows_NT-Release-arm as the test runtime folder.
Executing in C:\Users\robox\j\workspace\arm_cross_che---6fedd1e9\_\fx\bin\tests\System.Runtime.Tests\netcoreapp-Windows_NT-Release-arm\ 
----- start  0:15:29.22 ===============  To repro directly: ===================================================== 
pushd C:\Users\robox\j\workspace\arm_cross_che---6fedd1e9\_\fx\bin\tests\System.Runtime.Tests\netcoreapp-Windows_NT-Release-arm\
set COMPlus_JitStress=2
call C:\Users\robox\j\workspace\arm_cross_che---6fedd1e9\_\fx\bin\testhost\netcoreapp-Windows_NT-Release-arm\dotnet.exe xunit.console.netcore.exe System.Runtime.Tests.dll  -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapptests -notrait category=nonwindowstests  -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing
popd
===========================================================================================================

C:\Users\robox\j\workspace\arm_cross_che---6fedd1e9\_\fx\bin\tests\System.Runtime.Tests\netcoreapp-Windows_NT-Release-arm>set COMPlus_JitStress=2 

C:\Users\robox\j\workspace\arm_cross_che---6fedd1e9\_\fx\bin\tests\System.Runtime.Tests\netcoreapp-Windows_NT-Release-arm>call C:\Users\robox\j\workspace\arm_cross_che---6fedd1e9\_\fx\bin\testhost\netcoreapp-Windows_NT-Release-arm\dotnet.exe xunit.console.netcore.exe System.Runtime.Tests.dll  -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapptests -notrait category=nonwindowstests  -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
xUnit.net console test runner (32-bit .NET Core)
Copyright (C) 2014 Outercurve Foundation.

Discovering: System.Runtime.Tests
Discovered:  System.Runtime.Tests
Starting:    System.Runtime.Tests

Assert failure(PID 6980 [0x00001b44], Thread: 16036 [0x3ea4]): ExecutionManager::IsManagedCode(adrRedirectedIP)

<no module>! <no symbol> + 0x0 (0x00000000)
    File: d:\j\workspace\arm_cross_che---b006c545\src\vm\exceptionhandling.cpp Line: 5594
    Image: C:\Users\robox\j\workspace\arm_cross_che---6fedd1e9\_\fx\bin\testhost\netcoreapp-Windows_NT-Release-arm\dotnet.exe

----- end  0:16:09.48 ----- exit code 123456789 ----------------------------------------------------------
COREFX TEST FAILED
```

### PR DESCRIPTION
**Pull request by sandreenko**
Always add explicit null check before calling tail call stub that calls VSD.

Fixes #17585.

For arm64 it hits #6675.

### CODE REVIEW

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -7836,6 +7836,14 @@ void Compiler::fgMorphTailCall(GenTreeCall* call, void* pfnCopyArgs)
     JITDUMP("fgMorphTailCall (before):\n");
     DISPTREE(call);
 
+    // The runtime requires that we perforrm a null check on the `this` argument before tail
```
Comment (jkotas) [inline src/jit/morph.cpp]: Typo: perforrm

File: src/jit/morph.cpp [hunk 2]
```diff
@@ -7836,6 +7836,14 @@ void Compiler::fgMorphTailCall(GenTreeCall* call, void* pfnCopyArgs)
...
+    // calling to a virtual dispatch stub. This requirement is a consequence of limitations in the runtime's
+    // ability to map an AV to a NullReferenceException if the AV occurs in a dispatch stub.
```
Comment (jkotas) [inline src/jit/morph.cpp]: `... that has unmanaged caller`. The runtime is able to map an AV to a NullReferenceException exception in a general case. It is just the unmanaged caller case that is problematic.

File: src/jit/morph.cpp [hunk 3]
```diff
@@ -8205,7 +8207,12 @@ void Compiler::fgMorphTailCall(GenTreeCall* call, void* pfnCopyArgs)
     call->gtCallMoreFlags |= GTF_CALL_M_VARARGS | GTF_CALL_M_TAILCALL | GTF_CALL_M_TAILCALL_VIA_HELPER;
     call->gtFlags &= ~GTF_CALL_POP_ARGS;
 
-#endif // _TARGET_*
+#elif defined(_TARGET_ARM64_)
+    NYI("Tail calls via stub are unsupported on this platform.");
+#endif // _TARGET_ARM64_
+
+    // The function is responsible for doing explicit null check when it is neseccary.
```
Comment (jkotas) [inline src/jit/morph.cpp]: Typo: neseccary

File: src/jit/morph.cpp [hunk 4]
```diff
@@ -8205,7 +8207,12 @@ void Compiler::fgMorphTailCall(GenTreeCall* call, void* pfnCopyArgs)
     call->gtCallMoreFlags |= GTF_CALL_M_VARARGS | GTF_CALL_M_TAILCALL | GTF_CALL_M_TAILCALL_VIA_HELPER;
     call->gtFlags &= ~GTF_CALL_POP_ARGS;
 
-#endif // _TARGET_*
+#elif defined(_TARGET_ARM64_)
+    NYI("Tail calls via stub are unsupported on this platform.");
```
Comment (jkotas) [inline src/jit/morph.cpp]: These ARM64 specific NYIs do not hurt, but I would expect them to be required. Why are they not taken care of by `getTailCallCopyArgsThunk` returning null?

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib

File: src/jit/morph.cpp (hunk 4 above)
Comment (briansull) [reply src/jit/morph.cpp]: There also is  NYI_ARM64()  which might be better to use here.

File: src/jit/morph.cpp [hunk 5]
```diff
@@ -8733,6 +8740,11 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
                     // NYI - TAILCALL_RECURSIVE/TAILCALL_HELPER.
                     // So, bail out if we can't make fast tail call.
                     szFailReason = "Non-qualified fast tail call";
+
+                    if (call->IsTailPrefixedCall())
+                    {
+                        NYI("Arm64 does not support tail calls via helpers.");
```
Comment (briansull) [inline src/jit/morph.cpp]: There also is NYI_ARM64() which might be better to use here.

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked normal Build and Test

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked normal Build and Test

**Pull request merged**